### PR TITLE
Report some read errors back to java code.

### PR DESCRIPTION
**Pull request by hiddenalpha**

### CODE REVIEW

File: src/main/cpp/_nix_based/jssc.cpp [hunk 1]
```diff
@@ -549,22 +549,51 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
...
+static int awaitReadReady(JNIEnv*env, jlong fd){
+    int err;
+    int numUnknownErrors = 0;
 #if HAVE_POLL == 0
     // Alternative impl using 'select' as 'poll' isn't available (or broken).
 
-    //assert(fd < FD_SETSIZE); // <- Might help when hunting SEGFAULTs.
+    if( fd >= FD_SETSIZE ){
+        jclass exClz = env->FindClass("java/lang/UnsupportedOperationException");
+        if( exClz != NULL ) env->ThrowNew(exClz, "Bad luck. 'select' cannot hanlde large fds.");
```
Comment (tresf) [inline src/main/cpp/_nix_based/jssc.cpp:564]:
```suggestion
        if( exClz != NULL ) env->ThrowNew(exClz, "Bad luck. 'select' cannot handle large fds.");
```

File: src/main/cpp/_nix_based/jssc.cpp [hunk 2]
```diff
@@ -580,17 +609,32 @@ static void awaitReadReady(JNIEnv*, jlong fd){
...
-            continue;
+        if( result < 0 ){
+            err = errno;
+            jclass exClz = NULL;
+            switch( err ){
+            case EINVAL:
+                exClz = env->FindClass("java/lang/IllegalArgumentException");
+                if( exClz != NULL ) env->ThrowNew(exClz, "EINVAL poll()");
+                static_assert(EINVAL > 0, "EINVAL > 0");
+                return -EINVAL;
```
Comment (tresf) [inline src/main/cpp/_nix_based/jssc.cpp:620]: Just code consistency with your other switch case.

```suggestion
                return -err;
```

File: src/main/cpp/_nix_based/jssc.cpp [hunk 3]
```diff
@@ -549,22 +549,51 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
...
         int result = select(fd + 1, &readFds, NULL, NULL, NULL);
-        if(result < 0){
-            // man select: On error, -1 is returned, and errno is set to indicate the error
-            // TODO: Maybe a candidate to raise a java exception. But won't do
-            //       yet for backward compatibility.
-            continue;
+        if( result < 0 ){
+            err = errno;
+            jclass exClz = NULL;
+            switch( err ){
```
Comment (tresf) [inline src/main/cpp/_nix_based/jssc.cpp:576]: The `case` statements should probably be indented to match the styling of the rest of the file.

File: src/test/java/jssc/SerialNativeInterfaceTest.java [hunk 4]
```diff
@@ -50,4 +58,14 @@ public void reportsWriteErrorsAsIOException() throws Exception {
         testTarget.writeBytes(fd, buf);
     }
 
+    @Test
+    public void throwsIllegalArgumentExceptionIfPortHandleIllegal() throws Exception {
+        try{
+            testTarget.readBytes(999, 42);
```
Comment (tresf) [inline src/test/java/jssc/SerialNativeInterfaceTest.java:64]: Should probably be skipped on Windows.

File: src/test/java/jssc/SerialNativeInterfaceTest.java [hunk 5]
```diff
@@ -50,4 +59,16 @@ public void reportsWriteErrorsAsIOException() throws Exception {
         testTarget.writeBytes(fd, buf);
     }
 
+    @Test
+    public void throwsIllegalArgumentExceptionIfPortHandleIllegal() throws Exception {
+        assumeFalse(SerialNativeInterface.getOsType() == SerialNativeInterface.OS_WINDOWS);
```
Comment (tresf) [inline src/test/java/jssc/SerialNativeInterfaceTest.java:64]: Now that you've implemented the exception in `windows/jssc.cpp`'s readBytes, should we let this test run?

**Comment by tresf** [comment]
For some reason the Intel Mac job keeps getting hung in CI.  I'm going to see if I can determine if it's related to this PR or not before merging.

**Comment by tresf** [comment]
The hang definitely appears to occur during `SerialNativeInterfaceTest`.... What's weird is the CI just completely hangs and all of the `mvn` logs vanish once it gets to that point, as if we're somehow conflicting with the pipe... I'll see if I can reproduce locally.

**Comment by tresf** [comment]
I can reproduce the hang locally.  Investigating.

**Comment by tresf** [comment]
@hiddenalpha although I can reproduce the deadlock, I'm not sure why it's happening...

* On MacOS, when the test `throwsIllegalArgumentExceptionIfPortHandleIllegal` runs, it hangs indefinitely.
* I tried to add some `printf` statements to the C++ code, but could not find a way to display them from Junit.
* I tried to break when `result == 0`, but it doesn't affect the deadlock.
* I tried to provide different `portHandle` values, but they either behave the same or fail the test.

We may decide to just skip this test for MacOS, but before we make that decision, I'd like to know why.  I can offer a Mac environment via SSH if needed.

**Comment by pietrygamat** [comment]
The hanging test on MacOS seems related to `poll` vs `select` switch that causes MacOS use different path. From what I could tell, it is not related to the C code changes - the same test on master would get stuck the same way. Defning `#define HAVE_POLL 1` for [__APPLE__](https://github.com/java-native/jssc/blob/0cd6aef0ddf5e8fed6a450906b2f8afa6404086b/src/main/cpp/_nix_based/jssc.cpp#L46) as well as for other OSes results in [a test pass](https://github.com/pietrygamat/jssc/actions/runs/6936752651/job/18869498291), but that gets us back this:
```
    // Seems as poll has some portability issues on OsX (Search for "poll" in
    // "https://cr.yp.to/docs/unixport.html"). So we only make use of poll on
    // all platforms except "__APPLE__".
    // If you want to force usage of 'poll', pass "-DHAVE_POLL=1" to gcc.
```
and [perhaps this](https://beesbuzz.biz/code/5739-The-problem-with-select-vs-poll) and maybe this as well #53.

File: src/test/java/jssc/SerialNativeInterfaceTest.java [hunk 6]
```diff
@@ -1,16 +1,24 @@
...
 import static org.junit.Assert.fail;
 
 public class SerialNativeInterfaceTest {
 
+    private SerialNativeInterface testTarget;
+
+    @Before
+    public void before(){
+        testTarget = new SerialNativeInterface();
+    }
```
Comment (pietrygamat) [inline src/test/java/jssc/SerialNativeInterfaceTest.java:21]: This should best be moved from `@Before` directly to `throwsIllegalArgumentExceptionIfPortHandleIllegal()` method - just as it's done in `reportsWriteErrorsAsIOException`. We have no need to rerun this block for each test in this class, if only one is ever going to make use of it.

**Review by tresf — APPROVED** [review]

**Pull request merged**